shared/src/service_registry.rs

```diff
@@ -0,0 +1,215 @@
+                .downcast_ref::<Arc<T>>()
+                .expect("incorrectly registered service")
+                .clone(),
+        )
+    }
+}
+
+pub trait ServiceInitializable: Sized {
+    fn init(registry: &ServiceRegistry) -> Self;
+}
+
+impl<T> ServiceInitializable for T
+where
+    T: Default + Sized,
+{
+    fn init(_: &ServiceRegistry) -> Self {
+        T::default()
+    }
+}
+
+pub trait ServiceLinkable<T: ?Sized> {
+    fn as_link(self: Arc<Self>) -> Arc<T>;
+}
+
+#[macro_export]
+macro_rules! impl_service_linkable {
+    ($impl:ident => $trait:ident) => {
+        impl ServiceLinkable<dyn $trait> for $impl {
+            fn as_link(self: Arc<Self>) -> Arc<dyn $trait> {
+                self
+            }
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        bad_token::{BadTokenDetecting, MockBadTokenDetecting, TokenQuality},
+        gas_price_estimation::FakeGasPriceEstimator,
+        price_estimate::{BaselinePriceEstimator, PriceEstimating},
+        sources::uniswap::pool_fetching::{MockPoolFetching, PoolFetching},
+    };
+    use ethcontract::H160;
+    use gas_estimation::GasPriceEstimating;
+    use model::order::OrderKind;
+    use std::collections::HashSet;
+
+    trait Fooable: Send + Sync {
+        fn foo(&self) -> i32;
+    }
+
+    struct Foo;
+    impl Fooable for Foo {
+        fn foo(&self) -> i32 {
+            42
+        }
+    }
+    impl ServiceInitializable for Foo {
+        fn init(_: &ServiceRegistry) -> Self {
+            Foo
+        }
+    }
+    impl_service_linkable!(Foo => Fooable);
+
+    struct Bar(Arc<dyn Fooable>);
+    impl Bar {
+        fn magic(&self) -> i32 {
+            (self.0.foo() * 191) / 6
+        }
+    }
+    impl ServiceInitializable for Bar {
+        fn init(registry: &ServiceRegistry) -> Self {
+            Bar(registry.get())
+        }
+    }
+
+    #[test]
+    fn can_register_and_get() {
+        let mut registry = ServiceRegistry::default();
+        registry
+            // Register a `Foo` implementation for a `Fooable` component.
+            .register_as::<Foo, dyn Fooable>()
+            // Register a concrete `Bar` component.
+            .register::<Bar>();
+
+        assert_eq!(registry.get::<dyn Fooable>().foo(), 42);
+        assert_eq!(registry.get::<Bar>().magic(), 1337);
+    }
+
+    impl_service_linkable!(MockPoolFetching => PoolFetching);
+    impl_service_linkable!(FakeGasPriceEstimator => GasPriceEstimating);
+    impl_service_linkable!(MockBadTokenDetecting => BadTokenDetecting);
+    impl_service_linkable!(BaselinePriceEstimator => PriceEstimating);
+
+    // This would represent the command line arguments.
+    struct Options {
+        base_tokens: HashSet<H160>,
+        native_token: H160,
+    }
+
+    impl ServiceInitializable for BaselinePriceEstimator {
+        fn init(registry: &ServiceRegistry) -> Self {
+            let options = registry.get::<Options>();
```

Comment:
I'm a bit confused as to how the registry already has these options to get. I see only the struct was defined just above...

Reply:
It gets added to the registry that we create the `BaselinePriceEstimator` for:
```
        registry
            .add(Options {
                base_tokens: Default::default(),
                native_token: Default::default(),
            })
```

Reply:
I would understand this line more if I had seen registry dot add before registry dot get...

Reply:
The registry is a hashmap of components by type ID. So when initializing the `BaselinePriceEstimator` it expects all its dependencies to already be added to the hashmap.

In the test function, we create the hashmap (adding all the required dependencies) and then use it to initialize the `BaselinePriceEstimator`. The order is just positioning of code - they are separate functions so their position in the file does not reflect the call order.